File: src/main/java/com/shapyfy/core/domain/model/Exercise.java
package com.shapyfy.core.domain.model;

import jakarta.persistence.*;

import java.util.UUID;

@Entity(name = "exercises")
@Table(name = "exercises")
public class Exercise {

    @EmbeddedId
    private ExerciseId id;

    @Column(name = "translation_key", nullable = false, unique = true)
    private String translationKey;

    protected Exercise() {
        // JPA requires a no-arg constructor
    }

    private Exercise(ExerciseId id, String translationKey) {
        this.id = id;
        this.translationKey = translationKey;
    }

    public ExerciseId getId() {
        return id;
    }

    public String getTranslationKey() {
        return translationKey;
    }

    public static Exercise of(ExerciseId id, String translationKey) {
        return new Exercise(id, translationKey);
    }

    public static Exercise from(String translationKey) {
        return new Exercise(ExerciseId.createNew(), translationKey);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Exercise exercise = (Exercise) o;
        return id != null && id.equals(exercise.id);
    }

    @Override
    public int hashCode() {
        return getClass().hashCode();
    }

    @Override
    public String toString() {
        return "Exercise{" +
                "id=" + id +
                ", translationKey='" + translationKey + '\'' +
                '}';
    }

    @Embeddable
    public static class ExerciseId {
        private UUID id;

        protected ExerciseId() {
            // JPA requires a no-arg constructor
        }

        private ExerciseId(UUID id) {
            this.id = id;
        }

        public UUID getId() {
            return id;
        }

        public static ExerciseId of(UUID id) {
            return new ExerciseId(id);
        }

        public static ExerciseId createNew() {
            return new ExerciseId(UUID.randomUUID());
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            ExerciseId that = (ExerciseId) o;
            return id != null && id.equals(that.id);
        }

        @Override
        public int hashCode() {
            return id != null ? id.hashCode() : 0;
        }

        @Override
        public String toString() {
            return "ExerciseId{" + "id=" + id + '}';
        }
    }
}

File: src/main/kotlin/com/shapyfy/core/domain/model/Exercise.kt
package com.shapyfy.core.domain.model

import java.time.Instant
import java.util.UUID

/**
 * Core domain entity representing an exercise.
 * Name is the canonical identifier (e.g., "squat", "bench_press")
 */
data class Exercise(
    val id: UUID,
    val name: String,
    val createdAt: Instant,
    val updatedAt: Instant?
) {
    companion object {
        fun new(name: String): Exercise {
            require(name.isNotBlank()) { "Exercise name cannot be blank" }
            return Exercise(
                id = UUID.randomUUID(),
                name = name,
                createdAt = Instant.now(),
                updatedAt = null
            )
        }
    }
}

data class ExerciseCreationCommand(
    val rawName: String,
    val preferredLanguage: Language
) {
    init {
        require(rawName.isNotBlank()) { "Exercise name cannot be blank" }
    }
}

data class ExerciseCreationResult(
    val id: UUID,
    val localizedName: String,
    val translationKey: TranslationKey
)

data class ExerciseSummary(
    val id: UUID,
    val localizedName: String,
    val translationKey: TranslationKey
)
